File: EMR_BMED.Backend/Services/UserService.cs
using Microsoft.EntityFrameworkCore;

using EMR_BMED.Backend.Exceptions;
using EMR_BMED.Backend.Models;

namespace EMR_BMED.Backend.Services
{
  public class UserService(DbService dbService)
  {
    public async Task<UserModel> GetOneAsync(Guid id)
    {
      UserModel user = await dbService.Users
        .FirstOrDefaultAsync(
          user => user.Id == id)
        ?? throw new KeyNotFoundException($"Can't find user with id={id}");

      return user;
    }

    public PatientModel[] SearchPatients(string query)
    {
      PatientModel[] patients = dbService.Users
        .OfType<PatientModel>()
        .AsEnumerable()
        .Where(patient =>
          new string[] {
            patient.Email,
            patient.Name,
            patient.Surname,
            patient.SocialNumber,
          }.Any(s =>
            s.Contains(query, StringComparison.CurrentCultureIgnoreCase)
          )
        ).Take(50)
        .ToArray();

      return patients;
    }

    public IEnumerable<PatientModel> GetAllPatients()
    {
      return dbService.Users.OfType<PatientModel>().ToList();
    }

    public async Task UpdateAsync(Guid id, PatientUpdateDto dto)
    {
      PatientModel user = (PatientModel)await UpdateGenericAsync(id, dto);

      if (dto.SocialNumber != null) { user.SocialNumber = dto.SocialNumber; }
      if (dto.Citizenship != null) { user.Citizenship = dto.Citizenship; }
      if (dto.Allergies != null) { user.Allergies = dto.Allergies; }
      if (dto.Intolerances != null) { user.Intolerances = dto.Intolerances; }
      if (dto.Conditions != null) { user.Conditions = dto.Conditions; }
      if (dto.Blood != null) { user.Blood = dto.Blood; }

      await dbService.SaveChangesAsync();
    }

    public async Task UpdateAsync(Guid id, DoctorUpdateDto dto)
    {
      DoctorModel user = (DoctorModel)await UpdateGenericAsync(id, dto);

      if (dto.Address != null) { user.Address = dto.Address; }
      if (dto.MedicalField != null) { user.MedicalField = dto.MedicalField; }

      await dbService.SaveChangesAsync();
    }

    public async Task DeleteAsync(Guid id)
    {
      UserModel user = await GetOneAsync(id);

      dbService.Remove(user);
      await dbService.SaveChangesAsync();
    }

    private async Task<UserModel> UpdateGenericAsync(Guid id, UserUpdateDto dto)
    {
      UserModel user = await GetOneAsync(id);

      if (dto.Password != null) { user.Password = BCrypt.Net.BCrypt.HashPassword(dto.Password); }
      if (dto.Email != null)
      {
        if (dbService.Users.Any(u => u.Email == dto.Email))
        {
          throw new EmailIsTakenException();
        }
        user.Email = dto.Email;
      }
      if (dto.Surname != null) { user.Surname = dto.Surname; }
      if (dto.Name != null) { user.Name = dto.Name; }
      if (dto.Gender != null) { user.Gender = dto.Gender; }
      if (dto.IsVerified != null) { user.IsVerified = dto.IsVerified; }
      if (dto.Phone != null) { user.Phone = dto.Phone; }

      return user;
    }
  }
}

File: EMR_BMED.Backend.JS/services/user.service.js
import bcrypt from "bcrypt";

import db from "../data/EMR_BMED.json" with { type: "json" };
import dbService from "./db.service";

const getOne = (id) => {
  const user = db.users.find((u) => u.id === id);

  if (!!user) {
    throw new Error(`Can't find user with id=${id}`);
  }
  return user;
};

const searchPatients = (query) => {
  const patients = db.users
    .filter(
      (p) =>
        p.isDoctor === false &&
        [`${p.name} ${p.surname}`, p.email, p.socialNumber].some((s) =>
          s.toLowerCase().includes(query.toLowerCase())
        )
    )
    .slice(0, 50);

  return patients;
};

const getAllPatients = () => {
  return db.users.filter((p) => p.isDoctor === false);
};

const updateOne = (id, dto) => {
  const user = getOne(id);

  if (!!dto.password) {
    user.password = bcrypt.hashSync(dto.password, 1);
  }
  if (!!dto.email) {
    if (db.users.some((u) => u.email === dto.email)) {
      throw new Error("Email is taken");
    }
    user.email = dto.email;
  }
  if (!!dto.name) {
    user.name = dto.name;
  }
  if (!!dto.surname) {
    user.surname = dto.surname;
  }
  if (!!dto.gender) {
    user.gender = dto.gender;
  }
  if (!!dto.phone) {
    user.phone = dto.phone;
  }

  if (user.isDoctor) {
    if (!!dto.address) {
      user.address = dto.address;
    }
    if (!!dto.medicalField) {
      user.medicalField = dto.medicalField;
    }
  } else {
    if (!!dto.socialNumber) {
      user.socialNumber = dto.socialNumber;
    }
    if (!!dto.citizenship) {
      user.citizenship = dto.citizenship;
    }
    if (!!dto.allergies) {
      user.allergies = dto.allergies;
    }
    if (!!dto.intolerances) {
      user.intolerances = dto.intolerances;
    }
    if (!!dto.conditions) {
      user.conditions = dto.conditions;
    }
    if (!!dto.blood) {
      user.blood = dto.blood;
    }
  }

  db.users = db.users.map((u) => {
    if (u.id !== id) {
      return u;
    } else {
      return user;
    }
  });
  dbService.saveChanges(db);

  return user;
};

const deleteOne = (id) => {
  getOne(id);

  db.users = db.users.filter((u) => u.id !== id);
  dbService.saveChanges(db);
};

export default { getOne, searchPatients, getAllPatients, updateOne, deleteOne };
